streams/src/test/java/org/apache/kafka/streams/integration/KStreamRepartitionIntegrationTest.java

```diff
@@ -0,0 +1,548 @@
+
+        builder.stream(inputTopic, Consumed.with(Serdes.Integer(), Serdes.String()))
+            .selectKey((key, value) -> key.toString())
+            .repartition(Repartitioned.with(Serdes.String(), Serdes.String()))
+            .to(outputTopic);
+
+        startStreams(builder);
+
+        validateReceivedMessages(
+            new StringDeserializer(),
+            new StringDeserializer(),
+            Arrays.asList(
+                new KeyValue<>("1", "A"),
+                new KeyValue<>("2", "B")
+            )
+        );
+
+        final String topology = builder.build().describe().toString();
+
+        assertEquals(1, countOccurrencesInTopology(topology, "Sink: .*-repartition"));
+    }
+
+    @Test
+    public void shouldGoThroughRebalancingCorrectly() throws ExecutionException, InterruptedException {
+        final String repartitionName = "rebalancing-test";
+        final long timestamp = System.currentTimeMillis();
+
+        sendEvents(
+            timestamp,
+            Arrays.asList(
+                new KeyValue<>(1, "A"),
+                new KeyValue<>(2, "B")
+            )
+        );
+
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        final Repartitioned<String, String> repartitioned = Repartitioned.<String, String>as(repartitionName)
+            .withKeySerde(Serdes.String())
+            .withValueSerde(Serdes.String())
+            .withNumberOfPartitions(2);
+
+        builder.stream(inputTopic, Consumed.with(Serdes.Integer(), Serdes.String()))
+            .selectKey((key, value) -> key.toString())
+            .repartition(repartitioned)
+            .groupByKey()
+            .count()
+            .toStream()
+            .to(outputTopic);
+
+        startStreams(builder);
+        final KafkaStreams kafkaStreamsToClose = startStreams(builder);
+
+        validateReceivedMessages(
+            new StringDeserializer(),
+            new LongDeserializer(),
+            Arrays.asList(
+                new KeyValue<>("1", 1L),
+                new KeyValue<>("2", 1L)
+            )
+        );
+
+        kafkaStreamsToClose.close();
+
+        sendEvents(
+            timestamp,
+            Arrays.asList(
+                new KeyValue<>(1, "C"),
+                new KeyValue<>(2, "D")
+            )
+        );
+
+        validateReceivedMessages(
+            new StringDeserializer(),
+            new LongDeserializer(),
+            Arrays.asList(
+                new KeyValue<>("1", 2L),
+                new KeyValue<>("2", 2L)
+            )
+        );
+
+        final String repartitionTopicName = toRepartitionTopicName(repartitionName);
+
+        assertTrue(topicExists(repartitionTopicName));
+        assertEquals(2, getNumberOfPartitionsForTopic(repartitionTopicName));
+    }
+
+    private int getNumberOfPartitionsForTopic(final String topic) throws ExecutionException, InterruptedException {
+        try (AdminClient adminClient = createAdminClient()) {
```

Comment:
variable can be final. I won't comment on final-able variables anymore. Do the tests pass? There should be a check that fails on variables that aren't final, but could be.

Reply:
interesting... yup, checkstyle tests pass. I think checkstyle don't cover `try with resources` usage. anyway, it should be final yes, will update this everywhere. thanks.

Reply:
Ah, yeah, there are some limitations to the linter. Thanks for taking care of it.